File: app_python/tests/test_runtime.py
from datetime import datetime
from core.runtime import set_start_time, get_uptime
import core.runtime


def test_set_start_time():
    set_start_time()
    assert core.runtime.START_TIME is not None
    assert isinstance(core.runtime.START_TIME, datetime)


def test_get_uptime_success():
    set_start_time()
    uptime = get_uptime()
    
    assert "seconds" in uptime
    assert "human" in uptime
    assert isinstance(uptime["seconds"], int)
    assert isinstance(uptime["human"], str)
    assert uptime["seconds"] >= 0
    assert "hours" in uptime["human"]
    assert "minutes" in uptime["human"]


def test_get_uptime_increases():
    set_start_time()
    import time
    
    uptime1 = get_uptime()
    time.sleep(1)
    uptime2 = get_uptime()
    
    assert uptime2["seconds"] >= uptime1["seconds"]

File: app_go/test_runtime.go
package main

import (
	"testing"
	"time"
)

// Tests verify runtime calculations, uptime formatting, and time format validation.
func TestGetRuntime(t *testing.T) {
	// Reset startTime for testing
	originalStartTime := startTime
	startTime = time.Now().Add(-5 * time.Second)
	defer func() { startTime = originalStartTime }()

	runtime := getRuntime()

	// Check uptime is approximately 5 seconds (allow 1 second tolerance)
	if runtime.UptimeSeconds < 4 || runtime.UptimeSeconds > 6 {
		t.Errorf("Expected uptime around 5 seconds, got %d", runtime.UptimeSeconds)
	}

	// Check timezone
	if runtime.Timezone != "UTC" {
		t.Errorf("Expected timezone 'UTC', got '%s'", runtime.Timezone)
	}

	// Check uptime human format
	if runtime.UptimeHuman == "" {
		t.Error("UptimeHuman should not be empty")
	}
}

func TestFormatUptime(t *testing.T) {
	tests := []struct {
		hours   int64
		minutes int64
		want    string
	}{
		{0, 0, "0 hours, 0 minutes"},
		{1, 30, "1 hours, 30 minutes"},
		{24, 0, "24 hours, 0 minutes"},
	}

	for _, tt := range tests {
		got := formatUptime(tt.hours, tt.minutes)
		if got != tt.want {
			t.Errorf("formatUptime(%d, %d) = %s, want %s", tt.hours, tt.minutes, got, tt.want)
		}
	}
}

func TestGetRuntimeCurrentTime(t *testing.T) {
	originalStartTime := startTime
	startTime = time.Now().Add(-10 * time.Minute)
	defer func() { startTime = originalStartTime }()

	runtime := getRuntime()

	// Verify current time is in RFC3339 format
	if runtime.CurrentTime == "" {
		t.Error("CurrentTime should not be empty")
	}

	// Try to parse the time to verify format
	_, err := time.Parse(time.RFC3339, runtime.CurrentTime)
	if err != nil {
		t.Errorf("CurrentTime should be in RFC3339 format, got error: %v", err)
	}
}
